### DISCUSSION TITLE
UnhandledPromiseRejectionWarning: Error: Cannot find module 'gl' from 'node_modules/@luma.gl/webgl/dist/es5/webgl-utils/webgl-types.js'

### DISCUSSION THREAD

**Discussion by kboul** [visgl/deck.gl#6327] [Q&A] [answered]
I am using 
"react": "17.0.2",
"@deck.gl/core": "7.1.11",
"@deck.gl/layers": "7.1.11",
"@deck.gl/react": "7.1.11",
"nebula.gl": "0.15.0",
 "react-map-gl": "5.0.12",
"@testing-library/react": "^11.2.5",

in a react application and I am receiving in the testing console when I run my tests:
```
(node:29516) UnhandledPromiseRejectionWarning: Error: Cannot find module 'gl' from 'node_modules/@luma.gl/webgl/dist/es5/webgl-utils/webgl-types.js'
    at Resolver.resolveModule (/home/kboul/Documents/Apps/earthnetviewer/esa_react/node_modules/jest-runtime/node_modules/jest-resolve/build/index.js:306:11)
    at Resolver._getVirtualMockPath (/home/kboul/Documents/Apps/earthnetviewer/esa_react/node_modules/jest-runtime/node_modules/jest-resolve/build/index.js:445:14)
    at Resolver._getAbsolutePath (/home/kboul/Documents/Apps/earthnetviewer/esa_react/node_modules/jest-runtime/node_modules/jest-resolve/build/index.js:431:14)
    at Resolver.getModuleID (/home/kboul/Documents/Apps/earthnetviewer/esa_react/node_modules/jest-runtime/node_modules/jest-resolve/build/index.js:404:31)
    at Runtime._shouldMock (/home/kboul/Documents/Apps/earthnetviewer/esa_react/node_modules/jest-runtime/build/index.js:1521:37)
```
It goes far far way and then it ends up lilke this:
```
(node:29516)). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 8)
 PASS  src/earthseismic/SeismicAdministration/SeismicAdministration.test.js (49.098 s)
  ● Console

    console.log
      deck: deck.gl 7.1.10 - set deck.log.priority=1 (or higher) to trace attribute updates

      at Object.<anonymous> (node_modules/@deck.gl/core/src/lib/init.js:38:3)
```
Now the question is: Is there any way to deal with this and make it dissapear? I have upgraded `deck.gl` in another branch to version `8.4.x` but still receiving the same error?

**Comment by ibgreen** [comment]
`gl` should not be included in the browser. Does this happen when you are running tests under Node.js? Have you tried installing `gl`?

**Comment by kboul** [comment]
Thank you for the answer. I don't have any `gl` package installed. Is it required for any of the packages I am using? What do you mean by running tests under node.js? I am running the tests as you would run any tests of a react project.
I am not familiar with gl package. Do you mean this one? https://www.npmjs.com/package/gl

**Reply by ibgreen** [reply]
Yes that is the package. 

WebGL is not available under Node.js. If you want to run deck.gl based tests in Node.js you need a WebGL environment. Either you can use puppeteer to run headless chromium browser, but we also use headless gl module if that is installed.

Alternatively, just comment out your deck.gl tests under Node.js and run them only in the browser.

**Comment by kboul** [comment]
Ok, I think I missed that detail that Jest runs in a node environment. Sorry about that. But currently I only have tests inside components that deck. gl is being used. What if  I want to test only UI elements and not the map itself? Should  I comment all the tests being included in a component that map is used? But this is not the way to go in my opinion.I tried installing gl package but I get several errors and the installation fails. Is there any way to suppress the gl error on the testing console?

**Reply by ibgreen** [reply]
It should be possible to import deck.gl in a node.js environment, as long as you don't try to instantiate the `Deck` class.

So if you can add some logic to your app to prevent instantiating `Deck` when running Node tests you can test the rest of your UI.

**Reply by kboul** [reply]
Exactly, for instance on the component to be tested inside `beforeEach`. Is there anything related in the documentation on how can I do that (preventing instantiating Deck inside node tests). I would be so grateful if you can point me somewhere because I think I did not manage to find anything. And this error occupies the whole testing console. Thanks a lot in advance for all the pointers.

**Comment by kboul** [comment, marked as answer]
Solved in this [thread](https://github.com/visgl/deck.gl/discussions/6363)
